store/hash.go

```diff
@@ -0,0 +1,187 @@
+	"fmt"
+	"hash"
+	"reflect"
+	"strings"
+)
+
+type HashAlgorithm interface {
+	// ValidateHashString checks that a given input hash string is valid
+	ValidateHashString(hashStr string) error
+	// NormalizeHashString, given an input hash string, returns the
+	//normalized hash string depending on the hash algorithm implementation
+	NormalizeHashString(hashStr string) (string, error)
+	// HashToHashString takes a hash.Hash and returns a hash string. The
+	// hash.Hash must be of the supported type.
+	HashToHashString(h hash.Hash) (string, error)
+	// NewHash returns a new hash.Hash of the supported hashAlgorithm type
+	NewHash() hash.Hash
+}
+
+// SHA512HashAlgorithm is a hash algorithm that uses sha512 sums
+// To ameliorate excessively long hash string, hash strings use only the first
+// half of a sha512 rather than the entire hash.
+type SHA512HashAlgorithm struct {
+	hashPrefix              string
+	lenHash                 int
+	lenSum                  int
+	lenHashString           int
+	lenNormalizedSum        int
+	lenNormalizedHashString int
+	minlenHashString        int
+}
+
+func NewSHA512HashAlgorithm() *SHA512HashAlgorithm {
+	hashPrefix := "sha512"
+	lenHash := sha512.Size  // raw byte size
+	lenSum := (lenHash) * 2 // in hex characters
+	lenHashString := len(hashPrefix) + 1 + lenSum
+	lenNormalizedSum := (lenHash / 2) * 2 // half length, in hex characters
+	lenNormalizedHashString := len(hashPrefix) + 1 + lenNormalizedSum
+	minlenHashString := len(hashPrefix) + 1 + 2 // at least sha512-aa
+
+	return &SHA512HashAlgorithm{
+		hashPrefix:              hashPrefix,
+		lenHash:                 lenHash,
+		lenSum:                  lenSum,
+		lenHashString:           lenHashString,
+		lenNormalizedSum:        lenNormalizedSum,
+		lenNormalizedHashString: lenNormalizedHashString,
+		minlenHashString:        minlenHashString,
+	}
+}
+
+func (ha *SHA512HashAlgorithm) ValidateHashString(hashStr string) error {
+	elems := strings.Split(hashStr, "-")
+	if len(elems) != 2 {
+		return fmt.Errorf("badly formatted hash string")
+	}
+	hashPrefix := elems[0]
+	if hashPrefix != ha.hashPrefix {
+		return fmt.Errorf("wrong hash string prefix %q", hashPrefix)
+	}
+	if len(hashStr) > ha.lenHashString {
+		return fmt.Errorf("hash string too long")
+	}
+	if len(hashStr) < ha.minlenHashString {
+		return fmt.Errorf("hash string too short")
+	}
+	return nil
+}
+
+// NormalizeHashString, given an input hashstring, returns the hash string
+// cutted to lenNormalizedHashString if it's longer or an error if the hash
+// string is not valid
+func (ha *SHA512HashAlgorithm) NormalizeHashString(hashStr string) (string, error) {
+	if err := ha.ValidateHashString(hashStr); err != nil {
+		return "", err
+	}
+	if len(hashStr) > ha.lenNormalizedHashString {
+		hashStr = hashStr[:ha.lenNormalizedHashString]
+	}
+	return hashStr, nil
+}
+
+func (ha *SHA512HashAlgorithm) HashToHashString(h hash.Hash) (string, error) {
+	if reflect.TypeOf(h) != reflect.TypeOf(sha512.New()) {
```

Comment:
It could be worth doing a small optimization of computing & storing (in `init` or such) the sha512 `reflect.Type` only once, since `sha512.New` does some buffer allocation iirc.

I don't think there's a simple way to solve this without reflection or additional types/wrapping/complexity, but a comment that `sha512.digest` is private would help I think.